internal/rbd/rbd_healer.go

```diff
@@ -166,10 +169,21 @@ func runVolumeHealer(ns *NodeServer, conf *util.Config) error {
 			continue
 		}
 
-		err = callNodeStageVolume(ns, c, pv, conf.StagingPath)
-		if err != nil {
-			util.ErrorLogMsg("callNodeStageVolume failed for VolID: %s, err: %v",
-				pv.Spec.PersistentVolumeSource.CSI.VolumeHandle, err)
+		wg.Add(1)
+		go func(wg *sync.WaitGroup, ns *NodeServer, c *k8s.Clientset, pv *v1.PersistentVolume, stagingPath string) {
+			defer wg.Done()
+			channel <- callNodeStageVolume(ns, c, pv, stagingPath)
+		}(&wg, ns, c, pv, conf.StagingPath)
+	}
+
+	go func() {
+		wg.Wait()
```

Comment:
Only waiting in the background? Should there not be a `wg.Wait()` blocking the execution of this function until all NodeStageVolume calls are done?

Reply:
The example at https://golang.org/pkg/sync/#WaitGroup looks simpler with a single `wg.Wait()`. As that is part of the official docs, I wonder where the `wg.Done()` in a go-routine is documented.

Reply:
Yeah, this works pretty well, lets live with it.
If we find any issue in the future, it would be easy to address in a separate PR.

Reply:
Where did you get the example from? "this works pretty well" is not convincing me it is correct. I think it is really strange to call `wg.Wait()` in a new go-routine, as that suggests to me that the rest of the function can continue with the `wg.Wait()` running in the background.

From what I understand at the moment, it is expected that all `callNodeStageVolume()` go-routines have finished before `runVolumeHealer()` returns. And I doubt that is currently the case.

Reply:
> Where did you get the example from? "this works pretty well" is not convincing me it is correct. I think it is really strange to call `wg.Wait()` in a new go-routine, as that suggests to me that the rest of the function can continue with the `wg.Wait()` running in the background.
> 

Sure. Here are my references:



1. https://stackoverflow.com/questions/59638898/channel-synchronisation-with-waitgroup-closing-channel-and-waitgroup 
1. https://www.oreilly.com/library/view/concurrency-in-go/9781491941294/ch04.html
1. https://blog.golang.org/pipelines


> From what I understand at the moment, it is expected that all `callNodeStageVolume()` go-routines have finished before `runVolumeHealer()` returns. And I doubt that is currently the case.

No, it works just like your expectation :-)

Reply:
Thank, now I get it. The next for-loop reads all channels, and blocks until all channels have been read. I'm not sure if closing the channel in the background is very beneficial, but it also does not seem to hurt.